### PR TITLE
Activity Post on Public Group to Private Group do not change the privacy of the post - Fixed

### LINKED ISSUE #2171: Activity Post on Public Group to Private Group do not change the privacy of the post, Still visible on the public
**Describe the bug**
Public post on a public group still visible on public after changing the Group privacy to private.

**To Reproduce**
- create new test group and set it to public
- add new posts in feed of the group
- change test group to private
- the post that was published when it was still a public group remains public even if the group is already changed to private
- all content of the group should be changed to private since the group is already private

**Screenshots**
[https://prnt.sc/104qne0](https://prnt.sc/104qne0)
[https://prnt.sc/104qptp](https://prnt.sc/104qptp)
[https://prnt.sc/104qs6u](https://prnt.sc/104qs6u)

**Support ticket links**
[https://secure.helpscout.net/conversation/1431510630/126249/](https://secure.helpscout.net/conversation/1431510630/126249/)

### PR DESCRIPTION
**Pull request by mahedihasannoman**
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixes #2171  .

### How to test the changes in this Pull Request:

1. Create activity post on a public group
2. Change the group status to private
3. View as another user who is not a member of that group
4. See the activity feed
5. You will not see that group feed anymore

### Proof Screenshots or Video

https://www.loom.com/share/4716f0e88963454ab4f2201fb3503d44

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### CODE REVIEW

**Comment by chetansatasiya** [comment]
@ngolwalkar @krunal-buddyboss whoever gets the time first make sure to work on this as first priority. We need to make the sync below all cases.

1. Public to Private
2. Private to Public

**Comment by mahedihasannoman** [comment]
@chetansatasiya , @ngolwalkar & @krunal-buddyboss I have done the following sync
- Public to Private
- Private to Public
Please review now.

File: src/bp-groups/bp-groups-functions.php [hunk 1]
```diff
@@ -2652,6 +2668,41 @@ function groups_get_membership_requested_user_ids( $group_id = 0 ) {
...
+ */
+function groups_change_all_activity_feed_status( $group_id = 0, $status = 0 ) {
+	global $wpdb, $bp;
+
+	//bail if no group_id
+	if ( $group_id === 0 ) {
+		return false;
+	}
+
+	$q = $wpdb->prepare( "UPDATE {$bp->activity->table_name} SET hide_sitewide = %d WHERE item_id = %d and component = 'groups'", $status, $group_id );
```
Comment (chetansatasiya) [inline src/bp-groups/bp-groups-functions.php:2689]: @mahedihasannoman We need to add the check here like the privacy = 'public'
```suggestion
	$q = $wpdb->prepare( "UPDATE {$bp->activity->table_name} SET hide_sitewide = %d WHERE item_id = %d and component = 'groups' AND privacy = 'public' ", $status, $group_id );
```

**Review by chetansatasiya — CHANGES_REQUESTED** [review]
@mahedihasannoman Needs to update condition otherwise, it will break the document, video & photos data for a group.

**Review by chetansatasiya — APPROVED** [review]

**Comment by sumitbuddyboss** [comment]
Hello @mahedihasannoman 

Issue >
Still, I'm able to access all the private group activity in the news feed, So this fix is not working for me, please check the issue video.


Issue video:
https://www.loom.com/share/cd4d907ecea94557a62df034f1eaa458

cc @chetansatasiya 

Thank you

**Comment by mahedihasannoman** [comment]
Hi @sumitbuddyboss , I have also checked the issue but unable to replicate on my end unfortunately. Will you please see my video here: https://www.loom.com/share/d3a32d2a3a2f4261a46acaed61f55345
Will you please double check and confirm how we can replicate it.
Thank you.

**Comment by krunal-buddyboss** [comment]
@sumitbuddyboss 
As we have checked together we can not reproduce this issue anymore. please confirm from your side once more.
Thanks.
@mahedihasannoman

File: src/bp-groups/bp-groups-functions.php [hunk 2]
```diff
@@ -373,6 +373,22 @@ function groups_edit_group_settings( $group_id, $enable_forum, $status, $invite_
...
+	 */
+	if ( 'public' === $group->status && 'private' === $status ) {
+		groups_change_all_activity_feed_status( $group->id, 1 );
+	}
+
+	/**
+	 * Before we potentially switch the group status, if it has been changed to public
+	 * from private and there are existing activity feed, change feed status public. ( hide_sitewide = 0 )
+	 */
+	if ( 'private' === $group->status && 'public' === $status ) {
```
Comment (KartikSuthar) [inline src/bp-groups/bp-groups-functions.php:388]:
```suggestion
	if ( ( 'private' === $group->status || 'hidden' === $group->status ) && 'public' === $status ) {
```

**Review by KartikSuthar — CHANGES_REQUESTED** [review]
@krunal-buddyboss please do the following changes when group privacy updates to the public to hidden or hidden to the public.

**Comment by krunal-buddyboss** [comment]
Hello @sumitbuddyboss 

We have applied some changes to activity privacy when group privacy updates to the public to hidden or hidden to the public and also private to hidden or hidden to the private. So please test and confirm from your side as well.
Also merged the same to dev as well.
Thanks.
@KartikSuthar @chetansatasiya @sourabhmatolia @mahedihasannoman

**Review by KartikSuthar — APPROVED** [review]

**Comment by sumitbuddyboss** [comment]
Hello @krunal-buddyboss 

Fixed >
I have test fix/2171 and dev branch, now the group privacy functionality is working as expected. 
Checked the scenario like
 Public group to private 
 Private to Hidden
 Hidden to Public
Also check the subgroup, for proof please check the below video.

Video:
https://www.loom.com/share/f24371f718b24c40b6961fcd3aa65f61

Ready for release 🚀 

cc @chetansatasiya 

Thank you

**Comment by JetalDobariya** [comment]
Hello 

I have tested this PR in QA Approved branch and this PR's fictionalities are working fine related to Privacy of the groups.
Here, I attached Video for reference.
https://www.loom.com/share/bdea8a69b84f4cc7a9fbdf5638798702

So this PR is Ready for Productions.

Thanks!

**Pull request merged**